File: ios/EthosProtocol/Sources/Services/LogRedactor.swift
import Foundation

/// Utility for scrubbing sensitive values from strings and header dictionaries
/// before they are written to any diagnostic channel (os_log, print, crash
/// reports, `DecodingFailureLogger`, etc.).
///
/// This is a caseless enum (used as a namespace) so it cannot be instantiated.
enum LogRedactor {

    // MARK: - Sensitive header names (all lowercase for case-insensitive matching)

    /// HTTP header names whose values must always be replaced with `[REDACTED]`
    /// before logging. Matching is case-insensitive.
    static let sensitiveHeaders: Set<String> = [
        "authorization",
        "x-nonce",
        "x-otp",
        "x-2fa-token"
    ]

    // MARK: - Header redaction

    /// Returns a copy of `headers` where every key whose lowercased form is in
    /// `sensitiveHeaders` has its value replaced with `"[REDACTED]"`. All other
    /// headers are left unchanged.
    static func redactHeaders(_ headers: [String: String]) -> [String: String] {
        headers.mapValues { _ in "" } // placeholder; real impl below
        // Swift doesn't short-circuit mapValues for key access, so we use reduce:
        return headers.reduce(into: [String: String]()) { result, pair in
            let (key, value) = pair
            result[key] = sensitiveHeaders.contains(key.lowercased()) ? "[REDACTED]" : value
        }
    }

    // MARK: - String redaction

    /// Replaces known-sensitive patterns in `input` with safe placeholders:
    ///
    /// - `Bearer <token>` → `Bearer [REDACTED]`
    ///   Covers `Authorization: Bearer …` lines in logged request descriptions.
    ///
    /// - `x-nonce: <value>` → `x-nonce: [REDACTED]`  (case-insensitive, any delimiter)
    ///   Covers the anti-replay nonce header if it somehow appears in a log string.
    ///
    /// All replacements are performed with case-insensitive regex so the caller
    /// doesn't need to normalise casing first.
    static func redactString(_ input: String) -> String {
        var result = input

        // Redact Bearer tokens: "Bearer " followed by one or more token characters.
        // Token characters per RFC 6750: ALPHA / DIGIT / "-" / "." / "_" / "~" / "+" / "/"
        // plus base64url's "=". We use a broad character class to be conservative.
        if let bearerRegex = try? NSRegularExpression(
            pattern: #"Bearer\s+[A-Za-z0-9._\-~+/=]+"#,
            options: .caseInsensitive
        ) {
            let range = NSRange(result.startIndex..., in: result)
            result = bearerRegex.stringByReplacingMatches(
                in: result, range: range, withTemplate: "Bearer [REDACTED]"
            )
        }

        // Redact nonce values: "x-nonce" followed by optional whitespace / ":" / "=" and a value.
        if let nonceRegex = try? NSRegularExpression(
            pattern: #"(x-nonce\s*[=:]\s*)[A-Za-z0-9._\-]+"#,
            options: .caseInsensitive
        ) {
            let range = NSRange(result.startIndex..., in: result)
            result = nonceRegex.stringByReplacingMatches(
                in: result, range: range, withTemplate: "$1[REDACTED]"
            )
        }

        return result
    }
}

File: android/app/src/main/java/com/ethosprotocol/security/LogRedactor.kt
package com.ethosprotocol.security

/**
 * Utility for scrubbing sensitive values from strings and header maps before
 * they are written to any diagnostic channel (Logcat, crash reporters, Ktor
 * [io.ktor.client.plugins.logging.Logging], etc.).
 *
 * All matching is case-insensitive so callers don't need to normalise header
 * names before passing them in.
 */
object LogRedactor {

    // -------------------------------------------------------------------------
    // Sensitive header names (lowercase for case-insensitive comparison)
    // -------------------------------------------------------------------------

    /**
     * HTTP header names whose values must always be replaced with `[REDACTED]`
     * before logging. Matching is case-insensitive.
     */
    val SENSITIVE_HEADERS: Set<String> = setOf(
        "authorization",
        "x-nonce",
        "x-otp",
        "x-2fa-token"
    )

    // -------------------------------------------------------------------------
    // Header redaction
    // -------------------------------------------------------------------------

    /**
     * Returns a copy of [headers] where every entry whose key (case-insensitively)
     * is in [SENSITIVE_HEADERS] has its value replaced with `"[REDACTED]"`. All
     * other entries are left unchanged.
     */
    fun redactHeaders(headers: Map<String, String>): Map<String, String> =
        headers.mapValues { (key, value) ->
            if (SENSITIVE_HEADERS.contains(key.lowercase())) "[REDACTED]" else value
        }

    // -------------------------------------------------------------------------
    // String redaction
    // -------------------------------------------------------------------------

    private val BEARER_REGEX = Regex(
        pattern = """Bearer\s+[\w.\-~+/=]+""",
        options = setOf(RegexOption.IGNORE_CASE)
    )

    private val NONCE_REGEX = Regex(
        pattern = """(x-nonce\s*[=:]\s*)[\w.\-]+""",
        options = setOf(RegexOption.IGNORE_CASE)
    )

    /**
     * Replaces known-sensitive patterns in [input] with safe placeholders:
     *
     * - `Bearer <token>` → `Bearer [REDACTED]`
     *   Covers `Authorization: Bearer …` lines appearing in logged request dumps.
     *
     * - `x-nonce: <value>` → `x-nonce: [REDACTED]`  (case-insensitive)
     *   Covers the anti-replay nonce header if it appears in a log string.
     *
     * @param input The raw string to sanitise.
     * @return A copy of [input] with sensitive patterns replaced.
     */
    fun redactString(input: String): String {
        var result = BEARER_REGEX.replace(input, "Bearer [REDACTED]")
        result = NONCE_REGEX.replace(result) { match ->
            "${match.groupValues[1]}[REDACTED]"
        }
        return result
    }
}
